### PR TITLE
Test undo

### PR DESCRIPTION
**Pull request by BigheartSmallbrain**
Added test cases for undocommand feature. Fixed minor bugs in code and corresponding helper classes and functions.

### CODE REVIEW

File: src/test/java/ezschedule/testutil/EventBuilder.java [hunk 1]
```diff
@@ -29,6 +33,15 @@ public EventBuilder() {
...
     }
+    /**
+     * Creates a {@code EventBuilder} with the given details.
+     */
+    public EventBuilder(String name, String date, String startTime, String endTime) {
+        this.name = new Name(name);
+        this.date = new Date(date);
+        this.startTime = new Time(startTime);
+        this.endTime = new Time(endTime);
+    }
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/testutil/EventBuilder.java:44]: Not necessary. No need to add more code to maintain.

The same thing can be achieved by calling
``` java
Event e = new EventBuilder()
                  .withName(name)
                  .withDate(date)
                  .withStartTime(startTime)
                  .withEndTime(endTime)
                  .build();
```

File: src/test/java/ezschedule/testutil/EventBuilder.java [hunk 2]
```diff
@@ -14,6 +14,10 @@ public class EventBuilder {
     public static final String DEFAULT_DATE = "2023-05-01";
     public static final String DEFAULT_START_TIME = "08:00";
     public static final String DEFAULT_END_TIME = "10:00";
+    public static final String EDIT_EVENT_NAME = "Graduation";
+    public static final String EDIT_EVENT_DATE = "2023-10-10";
+    public static final String EDIT_EVENT_START_TIME = "10:00";
+    public static final String EDIT_EVENT_END_TIME = "22:00";
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/testutil/EventBuilder.java:20]: This feels like it should not be here.

File: src/main/java/ezschedule/commons/core/index/Index.java [hunk 3]
```diff
@@ -22,6 +22,9 @@ private Index(int zeroBasedIndex) {
 
         this.zeroBasedIndex = zeroBasedIndex;
     }
+    public static Index initIndex(int zeroBasedIndex) {
+        return new Index(zeroBasedIndex);
+    }
```
Comment (lhy-hoyin) [inline src/main/java/ezschedule/commons/core/index/Index.java:27]: What is the purpose of this? I can't seem to find where in this PR is this being used.

**Review by lhy-hoyin — CHANGES_REQUESTED** [review]

File: src/main/java/ezschedule/commons/core/index/Index.java (hunk 3 above)
Comment (lhy-hoyin) [inline src/main/java/ezschedule/commons/core/index/Index.java:27]: This seem redundant, and might also cause confusion. I recommend using with one of the following method found below: `fromZeroBased` or `fromOneBased`

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        List<Event> lastShownList = model.getFilteredEventList();
+        Index targetIndex = Index.initIndex(new Random().nextInt(lastShownList.size()));
+        Command editCommand = new EditCommand(targetIndex, new EditEventDescriptor());
+        Event eventToEdit = lastShownList.get(targetIndex.getZeroBased());
+        EventBuilder eventBuilder = new EventBuilder();
+        eventBuilder.withName(EventBuilder.EDIT_EVENT_NAME);
+        eventBuilder.withDate(EventBuilder.EDIT_EVENT_DATE);
+        eventBuilder.withStartTime(EventBuilder.EDIT_EVENT_START_TIME);
+        eventBuilder.withEndTime(EventBuilder.EDIT_EVENT_END_TIME);
+        Event editedEvent = eventBuilder.build();
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java:50]: You can make use of method chaining here.

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        List<Index> targetIndexes = new ArrayList<>();
+        targetIndexes.add(targetIndex);
+        Event eventToDelete = lastShownList.get(targetIndex.getZeroBased());
+        Command deleteCommand = new DeleteCommand(targetIndexes);
+        model.deleteEvent(eventToDelete);
+        model.addRecentEvent(eventToDelete);
+        model.addRecentCommand(deleteCommand);
+        assertCommandSuccess(new UndoCommand(), model, String.format(UndoCommand.MESSAGE_UNDONE_SUCCESS,
+                        deleteCommand.commandWord()),
+                expectedModel);
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java:71]: Fix indentation issue?

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        case "week": ((RecurCommand) recurCommand).addEventPerWeek(model, eventToRecur);
+            break;
+        case "month": ((RecurCommand) recurCommand).addEventPerMonth(model, eventToRecur);
+            break;
+        default:
+            break;
+        }
+        assertCommandSuccess(new UndoCommand(), model, String.format(UndoCommand.MESSAGE_UNDONE_SUCCESS,
+                        recurCommand.commandWord()),
+                expectedModel);
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java:110]: Fix indentation here?

**Review by lhy-hoyin — CHANGES_REQUESTED** [review]

**Review by JrmCkh — APPROVED** [review]
LGTM.

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 7]
```diff
@@ -67,13 +71,13 @@ public void execute_undoDeleteCommand_success() throws CommandException {
         model.addRecentEvent(eventToDelete);
         model.addRecentCommand(deleteCommand);
         assertCommandSuccess(new UndoCommand(), model, String.format(UndoCommand.MESSAGE_UNDONE_SUCCESS,
-                        deleteCommand.commandWord()),
+                deleteCommand.commandWord()),
                 expectedModel);
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java:75]: These 2 can actually be placed in the same line.

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 8]
```diff
@@ -82,13 +86,13 @@ public void execute_undoMultipleDeletedEvents_success() {
         Command deleteCommand = new DeleteCommand(targetIndexes);
         model.addRecentCommand(deleteCommand);
         assertCommandSuccess(new UndoCommand(), model, String.format(UndoCommand.MESSAGE_UNDONE_SUCCESS,
-                        deleteCommand.commandWord()),
+                deleteCommand.commandWord()),
                 expectedModel);
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java]: These 2 can actually be placed in the same line.

File: src/test/java/ezschedule/logic/commands/UndoCommandTest.java [hunk 9]
```diff
@@ -106,7 +110,7 @@ public void execute_undoRecurCommand_success() throws CommandException {
             break;
         }
         assertCommandSuccess(new UndoCommand(), model, String.format(UndoCommand.MESSAGE_UNDONE_SUCCESS,
-                        recurCommand.commandWord()),
+                recurCommand.commandWord()),
                 expectedModel);
```
Comment (lhy-hoyin) [inline src/test/java/ezschedule/logic/commands/UndoCommandTest.java:114]: These 2 can actually be placed in the same line.

**Review by lhy-hoyin — APPROVED** [review]

**Pull request merged**